### PR TITLE
Fix-up error handling in GPU detection

### PR DESCRIPTION
**Pull request by jakirkham**
Supersedes PR ( https://github.com/rapidsai/cudf/pull/4925 )
Related to issue ( https://github.com/rapidsai/cudf/issues/627 )

This does some cleanup of error handling within GPU detection functions of cuDF.

First this types CUDA C APIs as either returning `cudaError_t` or `CUresult` depending on whether they are host or device APIs. As `CUresult` was not defined before this adds that definition as well. In addition this adds `CUDADriverError` (analogous to `CUDARuntimeError`) to handle some of the differences from these errors.

Second when receiving an error code in Cython, we make sure to type it correctly (as either `cudaError_t` or `CUresult`). This avoids converting it to a Python object midway through execution of that function. Instead we delay this conversion until we are sure a Python exception needs to be created.

Third we make sure we are raising `CUDARuntimeError` or `CUDADriverError` appropriate as determined by whether the function called was a runtime or driver API function.

Should hopefully make error handling a little smoother, cleaner, and more efficient. Also should delay any C/C++ to Python conversion until something actually goes wrong. Hopefully makes the working cases a bit quicker (like during `import cudf`).

### CODE REVIEW

File: python/cudf/cudf/_cuda/gpu.pyx [hunk 1]
```diff
@@ -20,12 +23,37 @@ from cudf._cuda.gpu cimport underlying_type_attribute as c_attr
...
+        cdef str msg = cudaGetErrorString(status).decode()
         super(CUDARuntimeError, self).__init__(
-            '%s: %s' % (name.decode(), msg.decode()))
+            '%s: %s' % (name, msg))
+
+    def __reduce__(self):
+        return (type(self), (self.status,))
+
+
+class CUDADriverError(RuntimeError):
```
Comment (kkraus14) [inline python/cudf/cudf/_cuda/gpu.pyx:37]: We should really split these between runtime and driver files for future maintainability / our sanity.

Comment (jakirkham) [reply python/cudf/cudf/_cuda/gpu.pyx:37]: Sure. How should would you propose we do that? Also what other things should move?

Comment (kkraus14) [reply python/cudf/cudf/_cuda/gpu.pyx:37]: For starters: CUDARuntimeError vs CUDADriverError, CUresult vs cudaError, etc. I think we should differentiate based on header files where we have a corresponding `pxd` for each header file (`cuda.h`, `cuda_runtime_api.h`).

Comment (kkraus14) [reply python/cudf/cudf/_cuda/gpu.pyx:37]: We can do this as a follow up. For now this looks good to go.

File: python/cudf/cudf/_cuda/gpu.pyx [hunk 2]
```diff
@@ -335,16 +363,16 @@ def deviceGetName(int device):
...
 
     cdef char[256] device_name
-    cdef int status = cuDeviceGetName(
-        device_name,
-        sizeof(device_name),
-        device
+    cdef CUresult status = cuDeviceGetName(
+      device_name,
+      sizeof(device_name),
+      device
```
Comment (kkraus14) [inline python/cudf/cudf/_cuda/gpu.pyx:374]:
```suggestion
        device_name,
        sizeof(device_name),
        device
```

File: python/cudf/cudf/_cuda/gpu.pyx (hunk 1 above)
Comment (jakirkham) [reply python/cudf/cudf/_cuda/gpu.pyx:36]: Thanks Keith! Filed as issue ( https://github.com/rapidsai/cudf/issues/4961 ).

**Pull request merged**